File: src/org/mtransit/android/ui/MainActivity.java
package org.mtransit.android.ui;

import java.util.WeakHashMap;

import org.mtransit.android.R;
import org.mtransit.android.billing.IBillingManager;
import org.mtransit.android.commons.MTLog;
import org.mtransit.android.data.DataSourceProvider;
import org.mtransit.android.di.Injection;
import org.mtransit.android.ui.fragment.ABFragment;
import org.mtransit.android.ui.fragment.SearchFragment;
import org.mtransit.android.ui.view.common.IActivity;
import org.mtransit.android.util.AdsUtils;
import org.mtransit.android.util.AnalyticsUtils;
import org.mtransit.android.util.FragmentUtils;
import org.mtransit.android.util.MapUtils;

import android.app.Activity;
import android.app.SearchManager;
import android.content.Context;
import android.content.Intent;
import android.content.res.Configuration;
import android.location.Location;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.FrameLayout;

public class MainActivity extends MTActivityWithLocation implements
		FragmentManager.OnBackStackChangedListener,
		AnalyticsUtils.Trackable,
		IBillingManager.OnBillingResultListener,
		IActivity,
		DataSourceProvider.ModulesUpdateListener {

	private static final String TAG = "Stack-" + MainActivity.class.getSimpleName();

	@Override
	public String getLogTag() {
		return TAG;
	}

	private static final String TRACKING_SCREEN_NAME = "Main";

	@Override
	public String getScreenName() {
		return TRACKING_SCREEN_NAME;
	}

	public static Intent newInstance(Context context) {
		return new Intent(context, MainActivity.class);
	}

	private NavigationDrawerController navigationDrawerController;

	private ActionBarController abController;

	@NonNull
	private final IBillingManager billingManager;

	public MainActivity() {
		super();
		this.billingManager = Injection.providesBillingManager();
	}

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_main);
		this.abController = new ActionBarController(this);
		this.navigationDrawerController = new NavigationDrawerController(this);
		this.navigationDrawerController.onCreate(savedInstanceState);
		getSupportFragmentManager().addOnBackStackChangedListener(this);
		DataSourceProvider.addModulesUpdateListener(this);
		MapUtils.fixScreenFlickering((FrameLayout) findViewById(R.id.content_frame));
	}

	private boolean modulesUpdated = false;

	@Override
	public void onModulesUpdated() {
		this.modulesUpdated = true;
		if (!this.resumed) {
			return;
		}
		AdsUtils.onModulesUpdated(this);
		this.modulesUpdated = false; // processed
	}

	@Override
	public void onBillingResult(@Nullable Boolean hasSubscription) {
		if (hasSubscription != null) {
			AdsUtils.setShowingAds(!hasSubscription, this);
		}
	}

	@Override
	public void onActivityResult(int requestCode, int resultCode, Intent data) {
		super.onActivityResult(requestCode, resultCode, data);
	}

	public ActionBarController getAbController() {
		return abController;
	}

	public NavigationDrawerController getNavigationDrawerController() {
		return navigationDrawerController;
	}

	@Override
	protected void onNewIntent(Intent intent) {
		super.onNewIntent(intent);
		setIntent(intent);
		processIntent(intent);
	}

	private boolean processIntent(Intent intent) {
		if (Intent.ACTION_SEARCH.equals(intent.getAction())) {
			onSearchQueryRequested(intent.getStringExtra(SearchManager.QUERY));
			return true; // intent processed
		}
		return false; // intent not processed
	}

	@Override
	public boolean onSearchRequested() {
		onSearchQueryRequested(null);
		return true; // processed
	}

	public void onSearchQueryRequested(String query) {
		Fragment currentFragment = getCurrentFragment();
		if (currentFragment instanceof SearchFragment) {
			((SearchFragment) currentFragment).setSearchQuery(query, false);
		} else {
			addFragmentToStack(SearchFragment.newInstance(query, null, null), currentFragment);
		}
	}

	@Override
	protected void onStart() {
		super.onStart();
		if (this.navigationDrawerController != null) {
			this.navigationDrawerController.onStart();
		}
	}

	@Override
	protected void onResume() {
		super.onResume();
		if (this.abController != null) {
			this.abController.updateAB();
		}
		if (this.navigationDrawerController != null) {
			this.navigationDrawerController.onResume();
		}
		AnalyticsUtils.trackScreenView(this, this);
		this.billingManager.addListener(this);
		this.billingManager.refreshPurchases();
		AdsUtils.adaptToScreenSize(this, getResources().getConfiguration());
		onLastLocationChanged(getUserLocation());
	}

	@Override
	protected void onPostResume() {
		super.onPostResume();
		this.resumed = true;
		if (this.modulesUpdated) {
			new Handler().post(new Runnable() {
				@Override
				public void run() {
					if (MainActivity.this.modulesUpdated) {
						onModulesUpdated();
					}
				}
			});
		}
		DataSourceProvider.onResume();
	}

	private boolean resumed = false;

	public boolean isMTResumed() {
		return this.resumed;
	}

	@Override
	protected void onPause() {
		super.onPause();
		this.resumed = false;
		if (this.navigationDrawerController != null) {
			this.navigationDrawerController.onPause();
		}
		this.billingManager.removeListener(this);
		AdsUtils.pauseAd(this);
		DataSourceProvider.onPause();
	}

	@Override
	protected void onStop() {
		super.onStop();
		if (this.navigationDrawerController != null) {
			this.navigationDrawerController.onStop();
		}
	}

	@Override
	protected void onRestart() {
		super.onRestart();
		DataSourceProvider.resetIfNecessary(this);
		popFragmentsToPop();
	}

	@Override
	protected void onDestroy() {
		super.onDestroy();
		this.billingManager.destroy();
		DataSourceProvider.removeModulesUpdateListener(this);
		if (this.abController != null) {
			this.abController.destroy();
			this.abController = null;
		}
		if (this.navigationDrawerController != null) {
			this.navigationDrawerController.destroy();
			this.navigationDrawerController = null;
		}
		if (this.fragmentsToPopWR != null) {
			this.fragmentsToPopWR.clear();
			this.fragmentsToPopWR = null;
		}
		DataSourceProvider.destroy();
		AdsUtils.destroyAd(this);
	}

	@Override
	protected void onSaveInstanceState(Bundle outState) {
		if (this.navigationDrawerController != null) {
			this.navigationDrawerController.onSaveState(outState);
		}
		if (this.abController != null) {
			this.abController.onSaveState(outState);
		}
		super.onSaveInstanceState(outState);
	}

	@Override
	protected void onRestoreInstanceState(Bundle savedInstanceState) {
		super.onRestoreInstanceState(savedInstanceState);
		if (this.navigationDrawerController != null) {
			this.navigationDrawerController.onRestoreState(savedInstanceState);
		}
		if (this.abController != null) {
			this.abController.onRestoreState(savedInstanceState);
		}
	}

	public void clearFragmentBackStackImmediate() {
		FragmentUtils.clearFragmentBackStackImmediate(this, null);
	}

	public void showNewFragment(ABFragment newFragment, boolean addToStack, @Nullable Fragment optSource) {
		FragmentUtils.replaceFragment(this, R.id.content_frame, newFragment, addToStack, optSource);
		if (addToStack) {
			incBackEntryCount();
		}
		showContentFrameAsLoaded();
		if (this.abController != null) {
			FragmentUtils.executePendingTransactions(this, null);
			this.abController.setAB(newFragment);
			this.abController.updateABDrawerClosed();
		}
		if (this.navigationDrawerController != null) {
			this.navigationDrawerController.setCurrentSelectedItemChecked(getBackStackEntryCount() == 0);
		}
	}

	public void showContentFrameAsLoaded() {
		if (findViewById(R.id.content_frame_loading) != null) {
			findViewById(R.id.content_frame_loading).setVisibility(View.GONE);
		}
		if (findViewById(R.id.content_frame) != null) {
			findViewById(R.id.content_frame).setVisibility(View.VISIBLE);
		}
	}

	public void addFragmentToStack(ABFragment newFragment) {
		addFragmentToStack(newFragment, getCurrentFragment());
	}

	public void addFragmentToStack(ABFragment newFragment, @Nullable Fragment optSource) {
		showNewFragment(newFragment, true, optSource);
	}

	@Override
	public void onLastLocationChanged(@Nullable Location lastLocation) {
		MTActivityWithLocation.broadcastUserLocationChanged(this, getFragments(), lastLocation);
	}

	public boolean isCurrentFragmentVisible(Fragment fragment) {
		return FragmentUtils.isCurrentFragmentVisible(this, R.id.content_frame, fragment);
	}

	@Nullable
	private Fragment getCurrentFragment() {
		return FragmentUtils.getFragment(this, R.id.content_frame);
	}

	@Override
	public void onBackStackChanged() {
		resetBackStackEntryCount();
		if (this.abController != null) {
			this.abController.setAB((ABFragment) getCurrentFragment());
			this.abController.updateABDrawerClosed();
		}
		if (this.navigationDrawerController != null) {
			this.navigationDrawerController.onBackStackChanged(getBackStackEntryCount());
		}
	}

	@Override
	public void onBackPressed() {
		if (this.navigationDrawerController != null && this.navigationDrawerController.onBackPressed()) {
			return;
		}
		Fragment currentFragment = getCurrentFragment();
		if (currentFragment instanceof ABFragment && ((ABFragment) currentFragment).onBackPressed()) {
			return;
		}
		super.onBackPressed();
	}

	public void updateNavigationDrawerToggleIndicator() {
		if (this.navigationDrawerController != null) {
			this.navigationDrawerController.setDrawerToggleIndicatorEnabled(getBackStackEntryCount() < 1);
		}
	}

	public void enableNavigationDrawerToggleIndicator() {
		if (this.navigationDrawerController != null) {
			this.navigationDrawerController.setDrawerToggleIndicatorEnabled(true);
		}
	}

	public boolean isDrawerOpen() {
		return this.navigationDrawerController != null && this.navigationDrawerController.isDrawerOpen();
	}

	private Integer backStackEntryCount = null;

	public int getBackStackEntryCount() {
		if (this.backStackEntryCount == null) {
			initBackStackEntryCount();
		}
		return this.backStackEntryCount;
	}

	private void initBackStackEntryCount() {
		this.backStackEntryCount = getSupportFragmentManager().getBackStackEntryCount();
	}

	private void resetBackStackEntryCount() {
		this.backStackEntryCount = null;
	}

	private void incBackEntryCount() {
		if (this.backStackEntryCount == null) {
			initBackStackEntryCount();
		}
		this.backStackEntryCount++;
	}

	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		super.onCreateOptionsMenu(menu);
		if (this.abController != null) {
			this.abController.onCreateOptionsMenu(menu, getMenuInflater());
		}
		return true;
	}

	@Override
	protected void onPostCreate(Bundle savedInstanceState) {
		super.onPostCreate(savedInstanceState);
		if (this.navigationDrawerController != null) {
			this.navigationDrawerController.onActivityPostCreate();
		}
	}

	@Override
	public void onConfigurationChanged(Configuration newConfig) {
		super.onConfigurationChanged(newConfig);
		if (this.navigationDrawerController != null) {
			this.navigationDrawerController.onConfigurationChanged(newConfig);
		}
		AdsUtils.adaptToScreenSize(this, newConfig);
	}

	private WeakHashMap<Fragment, Object> fragmentsToPopWR = new WeakHashMap<Fragment, Object>();

	public void popFragmentFromStack(Fragment fragment) {
		FragmentUtils.popFragmentFromStack(this, fragment, null);
	}

	private void popFragmentsToPop() {
		try {
			if (this.fragmentsToPopWR != null) {
				for (Fragment fragment : this.fragmentsToPopWR.keySet()) {
					popFragmentFromStack(fragment);
				}
				this.fragmentsToPopWR.clear();
			}
		} catch (Exception e) {
			MTLog.w(this, e, "Error while poping fragments to pop from stack!");
		}
	}

	public boolean onUpIconClick() {
		return FragmentUtils.popLatestEntryFromStack(this, null);
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		if (this.navigationDrawerController != null && this.navigationDrawerController.onOptionsItemSelected(item)) {
			return true; // handled
		}
		if (this.abController != null && this.abController.onOptionsItemSelected(item)) {
			return true; // handled
		}
		return super.onOptionsItemSelected(item);
	}

	@NonNull
	@Override
	public Context getContext() {
		return this;
	}

	@NonNull
	@Override
	public Context requireContext() throws IllegalStateException {
		return this;
	}

	@NonNull
	@Override
	public Activity getActivity() {
		return this;
	}

	@NonNull
	@Override
	public Activity requireActivity() throws IllegalStateException {
		return this;
	}
}

File: src/main/java/org/mtransit/android/ui/main/MainActivity.kt
@file:JvmName("MainActivity") // ANALYTICS
package org.mtransit.android.ui.main

import android.annotation.SuppressLint
import android.app.PendingIntent
import android.app.SearchManager
import android.content.Context
import android.content.Intent
import android.content.res.Configuration
import android.graphics.drawable.ColorDrawable
import android.location.Location
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.activity.viewModels
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.lifecycle.lifecycleScope
import androidx.navigation.NavController
import androidx.navigation.fragment.NavHostFragment
import androidx.navigation.ui.AppBarConfiguration
import androidx.navigation.ui.NavigationUI
import androidx.navigation.ui.navigateUp
import androidx.navigation.ui.onNavDestinationSelected
import androidx.navigation.ui.setupActionBarWithNavController
import androidx.navigation.ui.setupWithNavController
import dagger.hilt.EntryPoint
import dagger.hilt.InstallIn
import dagger.hilt.android.AndroidEntryPoint
import dagger.hilt.android.EntryPointAccessors
import dagger.hilt.components.SingletonComponent
import kotlinx.coroutines.launch
import org.mtransit.android.R
import org.mtransit.android.ad.IAdManager
import org.mtransit.android.analytics.IAnalyticsManager
import org.mtransit.android.billing.IBillingManager
import org.mtransit.android.billing.IBillingManager.OnBillingResultListener
import org.mtransit.android.commons.LocaleUtils
import org.mtransit.android.commons.ThemeUtils
import org.mtransit.android.commons.registerReceiverCompat
import org.mtransit.android.databinding.ActivityMainBinding
import org.mtransit.android.datasource.DataSourcesRepository
import org.mtransit.android.dev.CrashReporter
import org.mtransit.android.dev.DemoModeManager
import org.mtransit.android.receiver.ModulesReceiver
import org.mtransit.android.task.ServiceUpdateLoader
import org.mtransit.android.task.StatusLoader
import org.mtransit.android.ui.MTActivityWithLocation
import org.mtransit.android.ui.search.SearchFragment
import org.mtransit.android.ui.setUpEdgeToEdge
import org.mtransit.android.ui.view.common.IActivity
import org.mtransit.android.util.NightModeUtils
import java.util.concurrent.atomic.AtomicInteger
import javax.inject.Inject

@Suppress("UNUSED_ANONYMOUS_PARAMETER", "unused", "MemberVisibilityCanBePrivate")
@AndroidEntryPoint
class MainActivity : MTActivityWithLocation(),
    FragmentManager.OnBackStackChangedListener,
    IAnalyticsManager.Trackable,
    OnBillingResultListener,
    IActivity,
    IAdManager.RewardedAdListener {

    companion object {
        private val LOG_TAG = "Stack2-" + MainActivity::class.java.simpleName
        private const val TRACKING_SCREEN_NAME = "Main"

        @JvmStatic
        fun newInstance(context: Context): Intent {
            return Intent(context, MainActivity::class.java)
        }
    }

    override fun getLogTag(): String = LOG_TAG

    override fun getScreenName() = TRACKING_SCREEN_NAME

    private val viewModel by viewModels<MainViewModel>()

    private lateinit var binding: ActivityMainBinding

    val navHostFragment: NavHostFragment
        get() = supportFragmentManager.findFragmentById(R.id.main_content) as NavHostFragment

    private lateinit var navController: NavController
    private lateinit var appBarConfig: AppBarConfiguration

    private var bgDrawable: ColorDrawable? = null
    private val abBgDrawable: ColorDrawable?
        get() {
            return bgDrawable ?: run {
                supportActionBar?.let { ab ->
                    ColorDrawable().apply {
                        bgDrawable = this
                        ab.setBackgroundDrawable(bgDrawable)
                    }
                }
                bgDrawable
            }
        }

    private val defaultBgColor by lazy { ThemeUtils.resolveColorAttribute(context, android.R.attr.colorPrimary) }

    @EntryPoint
    @InstallIn(SingletonComponent::class)
    interface MainActivityEntryPoint {
        val demoModeManager: DemoModeManager // used in attachBaseContext() before @Inject dependencies are available
    }

    private fun getEntryPoint(context: Context): MainActivityEntryPoint {
        return EntryPointAccessors.fromApplication(context.applicationContext, MainActivityEntryPoint::class.java)
    }

    override fun attachBaseContext(newBase: Context) {
        val demoModeManager = getEntryPoint(newBase).demoModeManager
        val fixedBase = if (demoModeManager.isForceLang()) {
            demoModeManager.fixLocale(newBase)
        } else {
            LocaleUtils.attachBaseContextActivity(newBase)
        }
        super.attachBaseContext(fixedBase)
        LocaleUtils.attachBaseContextActivityAfter(this)
    }

    @Inject
    lateinit var adManager: IAdManager

    @Inject
    lateinit var analyticsManager: IAnalyticsManager

    @Inject
    lateinit var crashReporter: CrashReporter

    @Inject
    lateinit var billingManager: IBillingManager

    @Inject
    lateinit var dataSourcesRepository: DataSourcesRepository

    @Inject
    lateinit var statusLoader: StatusLoader

    @Inject
    lateinit var serviceUpdateLoader: ServiceUpdateLoader

    @Inject
    lateinit var demoModeManager: DemoModeManager

    private var currentUiMode = -1
    override fun onCreate(savedInstanceState: Bundle?) {
        setUpEdgeToEdge()
        super.onCreate(savedInstanceState)
        adManager.init(this)
        NightModeUtils.resetColorCache() // single activity, no cache can be trusted to be from the right theme
        currentUiMode = resources.configuration.uiMode
        LocaleUtils.onCreateActivity(this)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        setSupportActionBar(binding.abToolbar)
        navController = navHostFragment.navController
        appBarConfig = AppBarConfiguration(
            viewModel.getRootScreenResId(),
            binding.drawerLayout
        )
        binding.navView?.setupWithNavController(navController)
        binding.bottomNavView?.setupWithNavController(navController) // 1st
        binding.bottomNavView?.setOnItemReselectedListener { item -> // 2nd: override to avoid re-selection
            viewModel.onItemReselected()
        }
        // TODO binding.navRailView?.setupWithNavController(navController) // 1st // required Navigation UI 2.4.0 stable https://developer.android.com/jetpack/androidx/releases/navigation
        binding.navRailView?.setOnItemSelectedListener { item -> // 1st
            viewModel.onItemSelected()
            NavigationUI.onNavDestinationSelected(item, navController)
        }
        binding.navRailView?.setOnItemReselectedListener { item -> // 2nd: override to avoid re-selection
            viewModel.onItemReselected()
        }
        setupActionBarWithNavController(navController, appBarConfig)
        navController.addOnDestinationChangedListener { controller, dest, args ->
            viewModel.onItemSelected()
            viewModel.onSelectedItemIdChanged(dest.id)
        }
        supportFragmentManager.addOnBackStackChangedListener(this)
        viewModel.selectedItemIdRes.observe(this) { selectedItemIdRes ->
            selectedItemIdRes?.let {
                val navGraph = navController.navInflater.inflate(R.navigation.nav_graph_main)
                navGraph.setStartDestination(selectedItemIdRes) // FIXME only when new value? ...
                navController.graph = navGraph
                showContentFrameAsLoaded()
            }
        }
        viewModel.showDrawerLearning.observe(this) { showDrawerLearning ->
            if (showDrawerLearning == true) {
                binding.drawerLayout?.let { drawerLayout ->
                    binding.navView?.let { navView ->
                        drawerLayout.openDrawer(navView)
                        viewModel.setUserLearnedDrawer(true)
                    }
                }
            }
        }
        viewModel.hasAgenciesEnabled.observe(this) { hasAgenciesEnabled ->
            // ad-manager does not persist activity but listen for changes itself
            adManager.onHasAgenciesEnabledUpdated(this, hasAgenciesEnabled)
        }
        viewModel.abTitle.observe(this) {
            binding.abToolbar.title = it
        }
        viewModel.abSubtitle.observe(this) {
            binding.abToolbar.subtitle = it
        }
        viewModel.abBgColor.observe(this) { newBgColor ->
            abBgDrawable?.color = newBgColor ?: defaultBgColor
        }
        billingManager.currentSubscription.observe(this) {
            // do nothing
        }
        registerReceiverCompat(ModulesReceiver(), ModulesReceiver.getIntentFilter(), ContextCompat.RECEIVER_NOT_EXPORTED)
    }

    override fun onBillingResult(productId: String?) {
        productId?.isNotEmpty()?.let { hasSubscription ->
            adManager.setShowingAds(!hasSubscription, this)
        }
    }

    override fun onNewIntent(@SuppressLint("UnknownNullness") intent: Intent) {
        super.onNewIntent(intent)
        setIntent(intent)
        processIntent(intent)
    }

    private fun processIntent(intent: Intent?): Boolean {
        if (Intent.ACTION_SEARCH == intent?.action) {
            onSearchQueryRequested(intent.getStringExtra(SearchManager.QUERY))
            return true // intent processed
        }
        return false // intent not processed
    }

    override fun onSearchRequested(): Boolean {
        onSearchQueryRequested(null)
        return true // processed
    }

    fun onSearchQueryRequested(query: String?) {
        (currentFragment as? SearchFragment)?.apply {
            setSearchQuery(query, false)
        } ?: run {
        }
    }

    override fun onResume() {
        super.onResume()
        adManager.adaptToScreenSize(this, resources.configuration)
        adManager.setRewardedAdListener(this) // used until POI screen is visible // need to pre-load ASAP
        adManager.linkRewardedAd(this)
        billingManager.addListener(this) // trigger onBillingResult() w/ current value
        billingManager.refreshPurchases()
        onLastLocationChanged(deviceLocation)
    }

    override fun onRewardedAdStatusChanged() {
        // DO NOTHING
    }

    override fun skipRewardedAd(): Boolean {
        return adManager.shouldSkipRewardedAd()
    }

    override fun onPostResume() {
        super.onPostResume()
        isMTResumed = true
        if (currentUiMode != resources.configuration.uiMode) {
            lifecycleScope.launch {
                NightModeUtils.setDefaultNightMode(activity, demoModeManager) // does NOT recreated because uiMode in configChanges AndroidManifest.xml
            }
        }
        viewModel.onAppVisible()
    }

    var isMTResumed = false
        private set

    override fun onPause() {
        super.onPause()
        isMTResumed = false
        billingManager.removeListener(this)
        adManager.pauseAd(this)
    }

    override fun onDestroy() {
        super.onDestroy()
        adManager.destroyAd(this)
        adManager.unlinkRewardedAd(this)
    }

    fun showContentFrameAsLoading() {
    }

    fun showContentFrameAsLoaded() {
    }

    override fun onLastLocationChanged(lastLocation: Location?) {
        broadcastDeviceLocationChanged(this, navHostFragment.childFragmentManager.fragments, lastLocation)
    }

    private var locationSettingsResolution: PendingIntent? = null

    override fun onLocationSettingsResolution(resolution: PendingIntent?) {
        this.locationSettingsResolution = resolution
        broadcastLocationSettingsResolutionChanged(this, navHostFragment.childFragmentManager.fragments, resolution)
    }

    override fun getLastLocationSettingsResolution() = this.locationSettingsResolution

    private val currentFragment: Fragment?
        get() = supportFragmentManager.primaryNavigationFragment // TODO ?

    override fun onBackStackChanged() {
        resetBackStackEntryCount()
    }

    override fun onSupportNavigateUp(): Boolean {
        return navController.navigateUp(appBarConfig) || super.onSupportNavigateUp()
    }

    fun updateNavigationDrawerToggleIndicator() {
    }

    fun enableNavigationDrawerToggleIndicator() {
    }

    private val _backStackEntryCount: AtomicInteger by lazy { AtomicInteger(supportFragmentManager.backStackEntryCount) }
    val backStackEntryCount: Int
        get() = _backStackEntryCount.get()

    private fun resetBackStackEntryCount() {
        _backStackEntryCount.set(supportFragmentManager.backStackEntryCount)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        super.onCreateOptionsMenu(menu)
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        if (this.currentUiMode != newConfig.uiMode) {
            NightModeUtils.setDefaultNightMode(context, demoModeManager) // does NOT recreated because uiMode in configChanges AndroidManifest.xml
            NightModeUtils.recreate(this) // not recreated because uiMode in configChanges AndroidManifest.xml
            return
        }
        adManager.adaptToScreenSize(this, newConfig)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return item.onNavDestinationSelected(navController) || super.onOptionsItemSelected(item)
    }
}